activerecord/lib/active_record/associations/association.rb

```diff
@@ -240,6 +241,17 @@ def raise_on_type_mismatch!(record)
           end
         end
 
+        def raise_on_klass_not_supporting_delegation
+          if !(klass.singleton_class < Delegation::DelegateCache)
```

Comment:
Is not better to check for the method that is needed? `klass.resond_to?(:relation_delegate_class)`

Reply:
Hello! I had this doubt, in fact :smile: 

The rationale is encapsulation - checking on on the module would treat that given method as an implementation detail (which is in fact used exclusively inside `DelegateCache`).

However, this was just a proposal, I'm a newbie to the AR internals - and I've separated this change into a commit for trivially reversing it.

Reply:
I think an inheritance check is better, because we have many API expectations and it's not useful to enumerate them, and seems arbitrary to just pick on whichever method is currently called first.

Thus, the check should be here: https://github.com/rails/rails/blob/4f0eb1ccd9652118b05d42f152adaa4df7dc3388/activerecord/lib/active_record/reflection.rb#L418-L420

And the error message should be `"#{klass} is not an Active Record model"`.

Reply:
:+1: for matthew suggestion

Reply:
Good news, everyone! Thanks to @matthewd's hint, the code is now much more streamlined - the new UT recycles existing classes, therefore I've dumped the one I introduced on the first iterations.

Ready for the (potentially) final review for me (with the condition of a green build).